File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    
    int N;
    boolean[] visited;
    ArrayList<Integer>[] adj;
    
    public int solution(int n, int[][] wires) {
        int answer = Integer.MAX_VALUE;
        
        N = n;
        adj = new ArrayList[N+1]; // 1 ~ N
        
        // wires의 i번째를 제외할 때 절댓값 구하고 비교
        for (int i = 0; i < n-1; i++) {
            
            visited = new boolean[N+1]; // 1 ~ N
            for (int j = 1; j <= n; j++) {
                adj[j] = new ArrayList<>();
            }
            
            
            for (int j = 0; j < n-1; j++) {
                if (i == j) continue; // i번째 제외
                
                int n1 = wires[j][0], n2 = wires[j][1];
                adj[n1].add(n2);
                adj[n2].add(n1);
            }
            
			int cnt1 = dfs(1);
            int cnt2 = 0;
            
            for (int j = 1; j <= N; j++) {
                if (!visited[j]) {
                    cnt2 = dfs(j);
                    break;
                }
            }
            
            answer = Math.min(answer, Math.abs(cnt1 - cnt2));
            
        }
        
        return answer;
    }
    
    private int dfs(int start) {
        visited[start] = true;
        int count = 1;
        
        for (int next: adj[start]) {
            if (visited[next]) continue;
			count += dfs(next);
        }
        
        return count;
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def solution(n, wires):
    """
    n: 송전탑 개수, 2 이상 100 이하
    wires: 전선 정보, n-1개
    
    - 모든 wire 없애는 경우에 대해 나눠진 2개 그래프 탐색
        - 시간복잡도: O(n * n) -> n이 100이므로 1만 이하, 0.1ms 이하 예상
    """
    
    # 인접 행렬 구성
    adj = [set() for _ in range(n)]
    for s, e in wires:
        adj[s - 1].add(e - 1)
        adj[e - 1].add(s - 1)        
    
    answer = 100
    
    def dfs(start):
        cnt = 1
        for nxt in adj[start]:
            if nxt in visited:
                continue
            visited.add(nxt)
            cnt += dfs(nxt)
        return cnt
    
    # 모든 간선 한번씩 없애보면서 그래프 탐색 2번씩 수행
    for s, e in wires:
        visited = set()
        visited.add(s - 1)
        visited.add(e - 1)
        adj[s - 1].remove(e - 1)
        adj[e - 1].remove(s - 1)
        answer = min(answer, abs(dfs(s - 1) - dfs(e - 1)))
        adj[s - 1].add(e - 1)
        adj[e - 1].add(s - 1)
        
    return answer